[c10d] Use c10::Synchronized in c10d::SequenceNum

dbort opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #74049

The new `c10::Synchronized` class helps enforce that locks are properly
held when touching guarded fields. Use it instead of a raw mutex.

Also add an apparently-missing TORCH_CHECK in get(). The header comment
says that it will throw if the number is not set, but get() didn't do
the check that the other methods did.

dbort committed:
[c10d] Use c10::Synchronized in c10d::SequenceNum

The new `c10::Synchronized` class helps enforce that locks are properly
held when touching guarded fields. Use it instead of a raw mutex.

Also add an apparently-missing TORCH_CHECK in get(). The header comment
says that it will throw if the number is not set, but get() didn't do
the check that the other methods did.

[ghstack-poisoned]

dbort commented:
@dhruvbird Here's a first pass based on your suggestion. I'm having a tough time finding a test target that covers this file, though.

TBH this construct is kind of awkward to me, especially for situations where multiple fields need to be guarded by the same mutex, which is the most common pattern I found. With `Synchronized`, I think we'd need to put all of those fields in a helper struct/class, which adds complexity and (slight) memory overhead/churn.

I think that investing in something like https://clang.llvm.org/docs/ThreadSafetyAnalysis.html may be more worthwhile.

dhruvbird commented on torch/csrc/distributed/c10d/sequence_num.cpp:
```diff
@@ -10,52 +10,61 @@ SequenceNum::SequenceNum() : num_(c10::nullopt) {}
-  } else {
-    num_ = other.get();
-  }
+  num_.withLock([&](auto& num) {
+    if (!other.isSet()) {
+      num = c10::nullopt;
+    } else {
+      num = other.get();
+    }
+  });
```
This is okay to do under a lock (not necessary), but there will be no contention since this the constructor. The new API mandates that this be done under a lock, so I wanted to highlight it.

dhruvbird commented on torch/csrc/distributed/c10d/sequence_num.cpp:
```diff
@@ -10,52 +10,61 @@ SequenceNum::SequenceNum() : num_(c10::nullopt) {}
-  std::lock_guard<std::mutex> lock(lock_);
-  return *num_;
+  uint64_t ret = 0;
+  num_.withLock([&](const auto num) {
+    TORCH_CHECK(num != c10::nullopt);
+    ret = *num;
+  });
+  return ret;
```
Once https://github.com/pytorch/pytorch/pull/73794/files lands you should be able to directly return the return value of the `withLock<T>()` method.

dhruvbird commented on torch/csrc/distributed/c10d/sequence_num.cpp:
```diff
@@ -10,52 +10,61 @@ SequenceNum::SequenceNum() : num_(c10::nullopt) {}
-  TORCH_CHECK(num_ != c10::nullopt);
-  curVal = *num_;
-  num_ = ++(*num_);
+  uint64_t curVal = 0;
+  num_.withLock([&](auto& num) {
+    TORCH_CHECK(num != c10::nullopt);
+    curVal = *num;
+    num = ++(*num);
+  });
   return curVal;
```
Same here (for return value).

dhruvbird reviewed (approved):
This seems fine! You can also send a follow up to address the comments regarding the return value once https://github.com/pytorch/pytorch/pull/73794/files lands (not necessary, but nice to have).

dhruvbird commented:
> @dhruvbird Here's a first pass based on your suggestion. I'm having a tough time finding a test target that covers this file, though.

From `fbsource/fbcode`, you can run: `buck query 'owner(caffe2/torch/csrc/distributed/c10d/sequence_num.cpp)'`

That tells you the rule. However, the graph doesn't provide the tests. IIRC, these are covered by the distributed tests in external (GitHub) CI, so if that passes, you're good to go!

> 
> TBH this construct is kind of awkward to me, especially for situations where multiple fields need to be guarded by the same mutex, which is the most common pattern I found. With `Synchronized`, I think we'd need to put all of those fields in a helper struct/class, which adds complexity and (slight) memory overhead/churn.

Yes, they need to be put into a separate class/struct. However, there's no additional memory overhead for doing so as long as the class has a member variable! :) 

> 
> I think that investing in something like https://clang.llvm.org/docs/ThreadSafetyAnalysis.html may be more worthwhile.

We do have some tsan runs set up internally, which detect these kinds of bugs, but they do rely on running the program :-/

dbort commented:
> > I think that investing in something like https://clang.llvm.org/docs/ThreadSafetyAnalysis.html may be more worthwhile.
> 
> We do have some tsan runs set up internally, which detect these kinds of bugs, but they do rely on running the program :-/ The link you pointed to does seem to be static (and doesn't require running the program), so we should definitely investigate the possibility of bringing that in!

Yep, it's static. I've used it on a bunch of projects, including low-level system runtimes and OS kernels, and it's really helpful and generally unobtrusive.

dbort committed:
Update on "[c10d] Use c10::Synchronized in c10d::SequenceNum"


The new `c10::Synchronized` class helps enforce that locks are properly
held when touching guarded fields. Use it instead of a raw mutex.

Also add an apparently-missing TORCH_CHECK in get(). The header comment
says that it will throw if the number is not set, but get() didn't do
the check that the other methods did.

[ghstack-poisoned]
